drivers/adc/adc_stm32.c

```diff
@@ -236,6 +236,41 @@ static const uint32_t table_samp_time[] = {
 };
 #endif
 
+/* table to define the LL ADC clock prescaler from the DTS */
+static const uint32_t adc_clock_prescaler[] = {
+#if defined(CONFIG_SOC_SERIES_STM32F0X) || \
+	defined(CONFIG_SOC_SERIES_STM32L0X) || \
+	defined(CONFIG_SOC_SERIES_STM32WLX)
+	LL_ADC_CLOCK_SYNC_PCLK_DIV1,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV2,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV4,
+#elif defined(STM32F3X_ADC_V1_1) || \
+	defined(CONFIG_SOC_SERIES_STM32L4X) || \
+	defined(CONFIG_SOC_SERIES_STM32L5X) || \
+	defined(CONFIG_SOC_SERIES_STM32WBX) || \
+	defined(CONFIG_SOC_SERIES_STM32G0X) || \
+	defined(CONFIG_SOC_SERIES_STM32G4X) || \
+	defined(CONFIG_SOC_SERIES_STM32H7X)
+	LL_ADC_CLOCK_SYNC_PCLK_DIV1,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV2,
+	LL_ADC_CLOCK_SYNC_PCLK_DIV4,
+#elif defined(CONFIG_SOC_SERIES_STM32L1X) || \
+	defined(CONFIG_SOC_SERIES_STM32U5X)
+	LL_ADC_CLOCK_ASYNC_DIV1,
+	LL_ADC_CLOCK_ASYNC_DIV2,
+	LL_ADC_CLOCK_ASYNC_DIV4,
+	LL_ADC_CLOCK_ASYNC_DIV6,
+	LL_ADC_CLOCK_ASYNC_DIV8,
+	LL_ADC_CLOCK_ASYNC_DIV10,
+	LL_ADC_CLOCK_ASYNC_DIV12,
+	LL_ADC_CLOCK_ASYNC_DIV16,
+	LL_ADC_CLOCK_ASYNC_DIV32,
+	LL_ADC_CLOCK_ASYNC_DIV64,
+	LL_ADC_CLOCK_ASYNC_DIV128,
+	LL_ADC_CLOCK_ASYNC_DIV256,
+#endif
+};
+
```

Comment:
```suggestion
/* table to define the LL ADC clock prescaler from the DTS */

#if defined(CONFIG_SOC_SERIES_STM32F0X) || \
	defined(CONFIG_SOC_SERIES_STM32L0X) || \
	defined(CONFIG_SOC_SERIES_STM32WLX)
        defined(STM32F3X_ADC_V1_1) || \
	defined(CONFIG_SOC_SERIES_STM32L4X) || \
	defined(CONFIG_SOC_SERIES_STM32L5X) || \
	defined(CONFIG_SOC_SERIES_STM32WBX) || \
	defined(CONFIG_SOC_SERIES_STM32G0X) || \
	defined(CONFIG_SOC_SERIES_STM32G4X) || \
	defined(CONFIG_SOC_SERIES_STM32H7X)
#define z_clk(v)	LL_ADC_CLOCK_SYNC_PCLK_##v
#elif defined(CONFIG_SOC_SERIES_STM32L1X) || \
	defined(CONFIG_SOC_SERIES_STM32U5X)
#define z_clk(v)	LL_ADC_CLOCK_ASYNC_##v
#endif

#define adc_clock_prescaler(v) z_clk(v)
```

Then you can do
```
LL_ADC_SetClock(adc, adc_clock_prescaler[config->clock_prescaler]);
```

Reply:
I think that specifying the actual clock-divider via dts instead of the register value would be more intuitive for the user.
And then you could use the macro recommended by erwango instead of using an array to translate the register value back and forth.

Reply:
Defining the st,prescaler property  directly with the divider value to be 1, 2, 4, 6, 8, 10, 12, 16, 32, 64,128,256  is possible in the dts. The related adc_clock_prescaler will depend on the soc serie 
either LL_ADC_CLOCK_SYNC_PCLK_##v or LL_ADC_CLOCK_ASYNC_##v 
Except for the **stm32F0** where default 1 is LL_ADC_CLOCK_ASYNC and 2 or 4 are LL_ADC_CLOCK_SYNC_PCLK_DIV_##v

Reply:
then configuring with  `LL_ADC_SetClock(adc, adc_clock_prescaler(config->clock_prescaler));`

Reply:
> However with` #define adc_clock_prescaler(v) z_clock_prescaler(v)` "v" must be like "DT_INST_PROP(0, st_prescaler)" and cannot be "config->clock_prescaler" (-->`error: 'LL_ADC_CLOCK_ASYNC_DIVconfig' undeclared`)

Right, since this rely on pre-processing macros 'v' should be available at build time, which is not the case for config->clock_prescaler.

Anyway, we should find another way to allow the user to provide the prescaler value instead of an array rank.

Reply:
Using COND_CODE_1, it is possible to build a LL_ADC_CLOCK_ASYNC_DIVn or LL_ADC_CLOCK_SYNC_PCLK_DIVn
value from the st,prescaler property (based on string concatenation)